### PR TITLE
Ensure that dag_run conf is a dict

### LINKED ISSUE #15023: DAG task execution and API fails if dag_run.conf is provided with an array or string (instead of dict)
**Apache Airflow version**: 2.0.1

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`): Tried both pip install and k8s image

**Environment**: Dev Workstation of K8s execution - both the same

- **OS** (e.g. from /etc/os-release): Ubuntu 20.04 LTS
- **Others**: Python 3.6

**What happened**:

We use Airflow 1.10.14 currently in production and have a couple of DAGs defined today which digest a batch call. We implemented the batch (currently) in a way that the jobs are provided as dag_run.conf as an array of dicts, e.g. "[ { "job": "1" }, { "job": "2" } ]".
Trying to upgrade to Airflow 2.0.1 we see that such calls are still possible to submit but all further actions are failing:
- It is not possible to query status via REST API, generates a HTTP 500
- DAG starts but all tasks fail.
- Logs can not be displayed (actually there are none produced on the file system)
- Error logging is a bit complex, Celery worker does not provide meaningful logs on console nor produces log files, running a scheduler as SequentialExecutor reveals at least one meaningful sack trace as below
- (probably a couple of other internal logic is also failing
- Note that the dag_run.conf can be seen as submitted (so is correctly received) in Browse--> DAG Runs menu

As a regression using the same dag and passing a dag_run.conf = "{ "batch": [ { "job": "1" }, { "job": "2" } ] }" as well as "{}".

Example (simple) DAG to reproduce:
```
from airflow import DAG
from airflow.operators.bash import BashOperator
from airflow.utils.dates import days_ago
from datetime import timedelta

dag = DAG(
    'test1',
    description='My first DAG',
    default_args={
        'owner': 'jscheffl',
        'email': ['***@***.de'],
        'email_on_failure': True,
        'email_on_retry': True,
        'retries': 5,
        'retry_delay': timedelta(minutes=5),
    },
    start_date=days_ago(2)
)

hello_world = BashOperator(
    task_id='hello_world',
    bash_command='echo hello world',
    dag=dag,
)
```

Stack trace from SequentialExecutor:
```
Traceback (most recent call last):
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/bin/airflow", line 8, in <module>
    sys.exit(main())
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/__main__.py", line 40, in main
    args.func(args)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/cli/cli_parser.py", line 48, in command
    return func(*args, **kwargs)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/utils/cli.py", line 89, in wrapper
    return f(*args, **kwargs)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/cli/commands/task_command.py", line 225, in task_run
    ti.init_run_context(raw=args.raw)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1987, in init_run_context
    self._set_context(self)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/utils/log/logging_mixin.py", line 54, in _set_context
    set_context(self.log, context)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/utils/log/logging_mixin.py", line 174, in set_context
    handler.set_context(value)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/utils/log/file_task_handler.py", line 56, in set_context
    local_loc = self._init_file(ti)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/utils/log/file_task_handler.py", line 245, in _init_file
    relative_path = self._render_filename(ti, ti.try_number)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/utils/log/file_task_handler.py", line 77, in _render_filename
    jinja_context = ti.get_template_context()
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/utils/session.py", line 65, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1606, in get_template_context
    self.overwrite_params_with_dag_run_conf(params=params, dag_run=dag_run)
  File "/home/jscheffl/Programmieren/Python/Airflow/syncignore/airflow-venv/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1743, in overwrite_params_with_dag_run_conf
    params.update(dag_run.conf)
ValueError: dictionary update sequence element #0 has length 4; 2 is required
{sequential_executor.py:66} ERROR - Failed to execute task Command '['airflow', 'tasks', 'run', 'test1', 'hello_world', '2021-03-25T22:22:36.732899+00:00', '--local', '--pool', 'default_pool', '--subdir', '/home/jscheffl/Programmieren/Python/Airflow/airflow-home/dags/test1.py']' returned non-zero exit status 1..
[2021-03-25 23:42:47,209] {scheduler_job.py:1199} INFO - Executor reports execution of test1.hello_world execution_date=2021-03-25 22:22:36.732899+00:00 exited with status failed for try_number 5
```

**What you expected to happen**:
- EITHER the submission of arrays as dag_run.conf is supported like in 1.10.14
- OR I would expect that the submission contains a validation if array values are not supported by Airflow (which it seems it was at least working in 1.10)

**How to reproduce it**: See DAG code above, reproduce the error e.g. by triggering with "[ "test" ]" as dag_run.conf

**Anything else we need to know**: I assume not :-)

### PR DESCRIPTION
**Pull request by jscheffl1**
With this PR new DAG triggers are validated that the submitted conf (=dag_run_conf) is a dictionary. Previously submissions were possible and the logic just validated if the parameter is a valid JSON. Nevertheless currently some business logic fails in Airflow if the con value is not a dictionary object.

Explicitly I added the validation NOT to the data model as instances with migrated legacy data (in Airflow 1.x non dictionary cases were supported) are not failing in loading data via ORM.

The PR mainly adds validation for Legacy Experimental API and WWW UI.
It adds pytest for Legacy Experimental API, "new" API and WWW UI.

Update Pylint and flynt

closes: #15023
related: #15023

### CODE REVIEW

File: airflow/providers/apache/hive/hooks/hive.py [hunk 1]
```diff
@@ -426,7 +426,7 @@ def load_file(
         if create or recreate:
             if field_dict is None:
                 raise ValueError("Must provide a field dict when creating a table")
-            fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
+            fields = ",\n    ".join([f"`{k.strip('`')}` {v}" for k, v in field_dict.items()])
```
Comment (ashb) [inline airflow/providers/apache/hive/hooks/hive.py]: This change looks unrelated -- is it?

Comment (ashb) [inline airflow/providers/apache/hive/hooks/hive.py]:
```suggestion
             fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
```

File: tests/providers/apache/hive/hooks/test_hive.py [hunk 2]
```diff
@@ -229,7 +229,7 @@ def test_load_file_create_table(self, mock_run_cli):
         filepath = "/path/to/input/file"
         table = "output_table"
         field_dict = OrderedDict([("name", "string"), ("gender", "string")])
-        fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
+        fields = ",\n    ".join([f"`{k.strip('`')}` {v}" for k, v in field_dict.items()])
```
Comment (ashb) [inline tests/providers/apache/hive/hooks/test_hive.py]:
```suggestion
         fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
```

File: airflow/providers/apache/hive/hooks/hive.py [hunk 3]
```diff
@@ -426,7 +426,7 @@ def load_file(
         if create or recreate:
             if field_dict is None:
                 raise ValueError("Must provide a field dict when creating a table")
-            fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
+             fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
```
Comment (ashb) [inline airflow/providers/apache/hive/hooks/hive.py]:
```suggestion
            fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
```

File: tests/providers/apache/hive/hooks/test_hive.py [hunk 4]
```diff
@@ -229,7 +229,7 @@ def test_load_file_create_table(self, mock_run_cli):
         filepath = "/path/to/input/file"
         table = "output_table"
         field_dict = OrderedDict([("name", "string"), ("gender", "string")])
-        fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
+         fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
```
Comment (ashb) [inline tests/providers/apache/hive/hooks/test_hive.py]:
```suggestion
        fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
```

**Review by ashb — APPROVED** [review]

File: airflow/providers/apache/hive/hooks/hive.py (hunk 1 above)
Comment (jscheffl1) [reply airflow/providers/apache/hive/hooks/hive.py]: Hi @ashb Yes, the changes in hive.py were unrelated but the test scripts failed on this preventing me to open the PR. I needed to fix there... don't know why :-(

Comment (ashb) [reply airflow/providers/apache/hive/hooks/hive.py]: It _should_ be ignored if you don't otherwise change the file.

**Comment by ashb** [comment]
@jscheffl Could you rebase this change please? (We split tests_views.py in to multiple files).

**Comment by jscheffl** [comment]
@ashb uups, something went bad in the rebase, will need one further attempt to integrate on rebase...

**Comment by jscheffl** [comment]
Ready for review (again)

File: airflow/providers/apache/hive/hooks/hive.py [hunk 5]
```diff
@@ -429,7 +429,7 @@ def load_file(
         if create or recreate:
             if field_dict is None:
                 raise ValueError("Must provide a field dict when creating a table")
-            fields = ",\n    ".join([f"`{k.strip('`')}` {v}" for k, v in field_dict.items()])
+            fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
```
Comment (ashb) [inline airflow/providers/apache/hive/hooks/hive.py]:
```suggestion
            fields = ",\n    ".join([f"`{k.strip('`')}` {v}" for k, v in field_dict.items()])
```

Should be able to do this

File: tests/providers/apache/hive/hooks/test_hive.py [hunk 6]
```diff
@@ -229,7 +229,7 @@ def test_load_file_create_table(self, mock_run_cli):
         filepath = "/path/to/input/file"
         table = "output_table"
         field_dict = OrderedDict([("name", "string"), ("gender", "string")])
-        fields = ",\n    ".join([f"`{k.strip('`')}` {v}" for k, v in field_dict.items()])
+        fields = ",\n    ".join(['`{k}` {v}'.format(k=k.strip('`'), v=v) for k, v in field_dict.items()])
```
Comment (ashb) [inline tests/providers/apache/hive/hooks/test_hive.py]:
```suggestion
        fields = ",\n    ".join([f"`{k.strip('`')}` {v}" for k, v in field_dict.items()])
```

File: tests/www/views/test_views_trigger_dag.py [hunk 7]
```diff
@@ -78,6 +78,16 @@ def test_trigger_dag_conf_malformed(admin_client):
     assert run is None
 
 
+def test_trigger_dag_conf_not_dict(self):
+    test_dag_id = "example_bash_operator"
+
+    response = self.client.post(f'trigger?dag_id={test_dag_id}', data={'conf': 'string and not a dict'})
+    self.check_content_in_response('must be a dict', response)
```
Comment (ashb) [inline tests/www/views/test_views_trigger_dag.py]:
```suggestion
    self.check_content_in_response('must be a dict', response, status_code=400)
```

**Comment by ashb** [comment]
```
tests/www/views/test_views_trigger_dag.py:87:30: F821 undefined name 'DR'
tests/www/views/test_views_trigger_dag.py:87:41: F821 undefined name 'DR'
```

File: tests/www/views/test_views_trigger_dag.py [hunk 8]
```diff
@@ -78,6 +78,16 @@ def test_trigger_dag_conf_malformed(admin_client):
...
+
+    run = self.session.query(DR).filter(DR.dag_id == test_dag_id).first()
```
Comment (ashb) [inline tests/www/views/test_views_trigger_dag.py]:
```suggestion
    run = self.session.query(DagRun).filter(DagRun.dag_id == test_dag_id).first()
```

File: tests/www/views/test_views_trigger_dag.py [hunk 9]
```diff
@@ -78,6 +78,16 @@ def test_trigger_dag_conf_malformed(admin_client):
     assert run is None
 
 
+def test_trigger_dag_conf_not_dict(self):
+    test_dag_id = "example_bash_operator"
+
+    response = self.client.post(f'trigger?dag_id={test_dag_id}', data={'conf': 'string and not a dict'})
+    self.check_content_in_response('must be a dict', response, status_code=400)
+
+    run = self.session.query(DagRun).filter(DagRun.dag_id == test_dag_id).first()
```
Comment (ashb) [inline tests/www/views/test_views_trigger_dag.py]:
```suggestion
def test_trigger_dag_conf_not_dict():
    test_dag_id = "example_bash_operator"

    response = admin_client.post(f'trigger?dag_id={test_dag_id}', data={'conf': 'string and not a dict'})
    check_content_in_response('must be a dict', response, status_code=400)

    with create_session() as session:
        run = session.query(DagRun).filter(DagRun.dag_id == test_dag_id).first()
```

**Comment by jscheffl** [comment]
@ashb Thanks for finalizing the merge! Looking forward for the next PR - I already have an idea what to contribute next... and hope I'll find some weekend time soon.

**Comment by ashb** [comment]
@jscheffl This PR will be in 2.1.1 too -- I merged it just before the freeze for release prep
